src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs

```diff
@@ -85,40 +122,118 @@ internal partial class RequestExecutionQueue
+                        this.Metrics.RecordCancellation();
+
+                        _completionSource.SetCanceled();
+                        return;
+                    }
+
+                    // If we weren't able to get a corresponding context for this request (for example, we
+                    // couldn't map a doc request to a particular Document, or we couldn't find an appropriate
+                    // Workspace for a global operation), then just immediately complete the request with a
+                    // 'null' response.  Note: the lsp spec was checked to ensure that 'null' is valid for all
+                    // the requests this could happen for.  However, this assumption may not hold in the future.
+                    // If that turns out to be the case, we could defer to the individual handler to decide
+                    // what to do.
+                    if (context == null)
+                    {
+                        _logger.TraceWarning($"Could not get request context for {MethodName}");
+                        this.Metrics.RecordFailure();
+
+                        _completionSource.SetResult(default);
+                        return;
+                    }
+
+                    var result = await _handler.HandleRequestAsync(_request, context.Value, cancellationToken).ConfigureAwait(false);
+
                     this.Metrics.RecordSuccess();
+
+                    _completionSource.SetResult(result);
                 }
-                catch (OperationCanceledException)
+                catch (OperationCanceledException ex)
                 {
+                    // Record logs + metrics on cancellation.
                     _logger.TraceInformation($"{MethodName} - Canceled");
                     this.Metrics.RecordCancellation();
-                    throw;
+
+                    // Only cancel the task completion source so the caller (client) can react.
+                    // We don't need cancellation exceptions bubbling up to the request queue.
+                    _completionSource.TrySetCanceled(ex.CancellationToken);
                 }
                 catch (Exception ex)
                 {
+                    // Record logs and metrics on the exception.
                     _logger.TraceException(ex);
                     this.Metrics.RecordFailure();
-                    throw;
+
+                    // Pass the exception to the task completion source, so the caller (client) can react
+                    _completionSource.SetException(ex);
+
+                    // Also allow the exception to flow back to the request queue to handle as appropriate
+                    throw new InvalidOperationException($"Error handling '{MethodName}' request: {ex.Message}", ex);
                 }
                 finally
                 {
                     _logger.TraceStop($"{MethodName} - Roslyn");
                 }
             }
+
+            public bool TrySetException(Exception e)
```

Comment:
nit.  `ex` or `exception`.   in general, please reserve single char names only for lambdas.